binary_search_tree/tree.py

```diff
@@ -1,59 +1,119 @@
+    def add(self, key, value=None):
+        new_tree_node = TreeNode(key, value)
+        if not self.root:
+            self.root = new_tree_node
+            return
+        return self.add_helper(self.root, new_tree_node)
+
+    def add_helper(self, current_node, new_node):
+        if current_node == None:
+            return new_node
+        else:
+            if new_node.key <= current_node.key:
+                current_node.left = self.add_helper(
+                    current_node.left, new_node)
+            else:
+                current_node.right = self.add_helper(
+                    current_node.right, new_node)
+            return current_node
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
 
-    # Time Complexity: 
-    # Space Complexity: 
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
+
+    def find_helper(self, current_node, key):
+        if current_node == None:
+            return None
+        elif current_node.key == key:
+            return current_node.value
+        elif key < current_node.key:
+            return self.find_helper(current_node.left, key)
+        return self.find_helper(current_node.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        ordered_list = []
+        self.inorder_helper(self.root, ordered_list)
+
+        return ordered_list
+
+    def inorder_helper(self, root, ordered_list):
+        if not root:
+            return
+        self.inorder_helper(root.left, ordered_list)
+        ordered_list.append({"key": root.key, "value": root.value})
+        self.inorder_helper(root.right, ordered_list)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
 
-    # Time Complexity: 
-    # Space Complexity:     
     def preorder(self):
-        pass
+        ordered_list = []
+        self.preorder_helper(self.root, ordered_list)
+        return ordered_list
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper(self, root, ordered_list):
+        if not root:
+            return
+        ordered_list.append({"key": root.key, "value": root.value})
+        self.preorder_helper(root.left, ordered_list)
+        self.preorder_helper(root.right, ordered_list)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        ordered_list = []
+        self.postorder_helper(self.root, ordered_list)
+        return ordered_list
+
+    def postorder_helper(self, root, ordered_list):
+        if not root:
+            return
+        self.postorder_helper(root.left, ordered_list)
+        self.postorder_helper(root.right, ordered_list)
+        ordered_list.append({"key": root.key, "value": root.value})
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
 
-    # Time Complexity: 
-    # Space Complexity:     
     def height(self):
```

Comment:
👍  But the space complexity is O(log n) for a balanced tree due to the call stack because of the recursive nature of this solution.  For an unbalanced tree this would be O(n) space complexity.